test/Closures/update-funcexpr.js

```diff
@@ -0,0 +1,57 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+if (this.WScript && this.WScript.LoadScriptFile) { // Check for running in ch
+    this.WScript.LoadScriptFile("..\\UnitTestFramework\\UnitTestFramework.js");
+}
+
+var tests = [
+    {
+        name: "Functions can overwrite themselves",
+        body: function () {
+            function foo1() {
+                foo1 = 42;
+                assert.isTrue(typeof foo1 == "number", "foo1 should overwrite itself to a number");
+                assert.areEqual(42, foo1, "value of foo1 after assingment");
+            }
+            foo1();
+
+            function foo2() {
+                foo2 &= 0;
+                assert.isTrue(typeof foo2 == "number", "foo2 should overwrite itself to a number");
+                assert.areEqual(0, foo2, "value of foo2 after assingment");
+            }
+            foo2();
+
+            function foo3() {
+                foo3 <<= 0;
+                assert.isTrue(typeof foo3 == "number", "foo3 should overwrite itself to a number");
+                assert.areEqual(0, foo3, "value of foo3 after assingment");
+            }
+            foo3();
+        }
+    },
+    {
+        name: "Function expressions cannot overwrite themselves",
+        body: function () {
+            (function foo1() {

```

Comment:
another interesting case here might be:

var foo = function foo() { … }

Reply:
You mean, to make sure the assignment to the var does happen?

Reply:
In that case the inner foo won't get overwritten (as this is testing) but the outer var does not change either since the reference in the body points to the function expression name rather than the var.

Reply:
(function foo() {
    var foo = foo;
    console.log(foo);
})();
// prints out 'undefined' (same in Chrome)

var bar = function bar() { console.log(bar); }
bar();
// prints out 'function bar() { console.log(bar); }' (same in Chrome)

So it seems that we are doing the right thing already (and I'd think it's a different scenario from attempting to assign to a function expression I'm fixing). Or did you have something else in mind?